skimage/morphology/_restorable_queue.pxi

```diff
@@ -0,0 +1,119 @@
+"""FIFO queue whose items can be iterated multiple times.
+
+.. warning:: Expects "QueueItem" to be defined before inclusion!
+
+This queue can be operated like a class. The structure stores the state of the
+instance and the contained functions act as instance methods. The important
+distinction compared to normal queues is that popping an element doesn't remove
+it internally. Thus unless the queue's internal buffer is explicitly cleared
+(see `queue_clear`) popped items can be restored at anytime using
+`queue_restore`.
+
+Example
+-------
+::
+
+    cdef RestorableQueue queue
+    cdef QueueItem item
+
+    queue_init(&queue, 5)
+    try:
+        for i in range(10):
+            item = <QueueItem>i
+            queue_push(&queue, &item)
+
+        while queue_pop(&queue, &item):
+            print(item)  # Prints 0 to 9
+
+        queue_restore(&queue)
+
+        while queue_pop(&queue, &item):
+            print(item)  # Prints 0 to 9 again
+    finally:
+        queue_exit(&queue)
+"""
+
+from libc.stdlib cimport malloc, realloc, free
+
+
+# Store state of queue
+cdef struct RestorableQueue:
+    QueueItem* _buffer_ptr
+    Py_ssize_t _buffer_size  # Maximal number of elements the buffer can store
+    Py_ssize_t _index_valid  # Index to most recently inserted item
+    Py_ssize_t _index_consumed  # Index to most recently consumed item
+
+
+cdef inline void queue_init(RestorableQueue* self, Py_ssize_t buffer_size) nogil:
+    """Initialize the queue and its buffer size.
+    
+    The size is defined as the number of queue items to fit into the initial 
+    buffer, thus its true memory size is `buffer_size * sizeof(QueueItem)`. Be
+    sure to call `queue_exit` after calling this function to free the allocated
+    memory!
+    """
+    self._buffer_ptr = <QueueItem*>malloc(buffer_size * sizeof(QueueItem))
+    if not self._buffer_ptr:
+        with gil:
+            raise MemoryError("couldn't allocate buffer")
+    self._buffer_size = buffer_size
+    self._index_consumed = -1
+    self._index_valid = -1
+
+
+cdef inline void queue_push(RestorableQueue* self, QueueItem* item_ptr) nogil:
```

Comment:
I'm certainly no expert either but I'll voice my thoughts on this anyway. Please keep that in mind. :)

My understanding is that the [GIL](https://wiki.python.org/moin/GlobalInterpreterLock) is meant to protect Python objects from being accessed multiple times at once from different threads. This would be dangerous because CPython's memory management is not thread-safe. 

However (concerning the current PR) the queue doesn't operate on objects accessible from Python. Furthermore `_local_maxima`'s queue is not shared between multiple threads, even if we would try to call this function multiple times in parallel from different Python threads (desirable because it releases the GIL and thus allows "true parallelism") each call would have its own queue only accessible within the algorithm and thus single thread. So I'd argue that this is safe as long as one ensures that access from multiple threads is currently not possible.

I remember a slightly related discussion about [Memoryviews and the GIL](https://github.com/scipy/scipy/pull/8541#discussion_r175613803) in one of my previous PRs. Might be useful to understand the issue.

Reply:
Ah, if you can't access the same instance of a Queue from Python, then all is well. Thanks for pointing that out! =)

With memoryviews, my (again, naive) understanding is that it's ok as long as you are only reading (not writing) and you know that no one is going to delete the memory from under you.

Reply:
> as long as you are only reading (not writing)

Well I'm certainly writing to the `flags` memoryview! But I still think this is safe because Cython's doc only mentions that the GIL is needed if

> [... the dtype is object and an object element is read or written.](http://docs.cython.org/en/latest/src/userguide/memoryviews.html#memoryviews-and-the-gil)

Reply:
:+1: